o2a/o2a_libs/el_parser.py

```diff
@@ -0,0 +1,351 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Expression Language to Jinja translator.
+"""
+
+__all__ = ["translate"]
+
+from typing import Union
+import re
+
+from lark import Lark, Tree, Token
+
+
+GRAMMAR = r"""
+    start: "${" lvalue_inner "}"
+        | "#{" lvalue_inner "}"
+        | ("${" expression "}")+
+        | ("#{" expression "}")+
+        | literal_expression
+
+    lvalue_inner: identifier
+        | non_litera_lvalue_prefix (value_suffix)*
+
+    literal_expression: literal_component (/[\$\#]/)?
+
+    // TODO: extend to full rule ! ! !
+    literal_component: (/[^\$\#]/)*
+        | (/[^\$\#]/)* (/[\$\#][^\{]/)
+        | (/[^\$\#]/)* /\\\\/ (/[\$\#]/)?
+
+    expression: expression1 ternary?
+
+    ternary: "?" expression ":" expression
+
+    expression1: expression binary_op expression
+        | unary_expression
+
+    binary_op: "and"
+        | "&&"
+        | "or"
+        | "||"
+        | "+"
+        | "-"
+        | "*"
+        | "/"
+        | "div"
+        | "%"
+        | "mod"
+        | ">"
+        | "gt"
+        | "<"
+        | "lt"
+        | ">="
+        | "ge"
+        | "<="
+        | "le"
+        | "=="
+        | "eq"
+        | "!="
+        | "ne"
+
+    unary_expression: unary_op unary_expression
+        | value
+
+    unary_op: "-"
+        | "!"
+        | "not"
+        | "empty"
+
+    value: value_prefix (value_suffix)*
+
+    value_prefix: literal
+        | non_litera_lvalue_prefix
+
+    non_litera_lvalue_prefix: "(" expression ")"
```

Comment:
should be non_literal_lvalue_prefix maybe ?

Reply:
You are right!